src/System.Data.Common/src/System/Data/Common/DbProviderFactory.netcoreapp.cs

```diff
@@ -0,0 +1,56 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Data.Common
+{
+    public abstract partial class DbProviderFactory
+    {
+        private bool? _canCreateDataAdapter;
+        private bool? _canCreateCommandBuilder;
+
+        public virtual bool CanCreateDataAdapter
+        {
+            get
+            {
+                if (!_canCreateDataAdapter.HasValue)
+                {
+                    var adapter = CreateDataAdapter();
```

Comment:
Is it expected that if this throws, then that exception will escape? Do we know if it is normal for this to return `null` rather than throwing if creation is not supported?

Reply:
The null check is primarily for the default implementation of `DbProviderFactory.CreateDataAdapter()` which simply returns null - it's expected that overriding implementations simply return an adapter from that method. I'm not sure what reasons a provider might have to throw an exception there, but I'm not convinced that one  should be interpreted as lack of support - it may rather indicate some sort of resource exhaustion, for example. If you feel it makes more sense to catch the exception and return false, I'll do that.

In terms of what is normal, we don't have much in the way of example implementations - AFAIK the only provider that doesn't support DataAdapter is Sqlite...

Reply:
Okay. If we do find providers that throw `NotSupportedException` or `NotImplementedException`, then I guess we will ask them to change their implementation to match the convention of returning null?

Reply:
I think it's OK for providers to throw NotSupportedException - in fact it's better than returning null... If a provider does throw NSE (or anything else), I'd ask them to override the new CanCreate* methods to just return false instead. That sound reasonable?

PS Originally I actually proposed changing the default implementation to throw NSE instead of returning null, but we didn't do it because we don't want to break backwards compat where users are already expecting null.

Reply:
@roji That seems reasonable.